Bugfix/get all record duplicated columns

lavigne958 opened the pull request:
Add a new argument to `get_all_records` to provide the list of expected
headers.

The given expected headers must:
- be unique
- be part of the complete headers list
- must not contain extra headers

This will provide a way for users to use this method and still
have *some* duplicated headers that are not relevant to pull.

This will ensure the columns that matters have unique headers.

Closes #1007

robrap commented on gspread/worksheet.py:
```diff
@@ -399,6 +400,7 @@ def get_all_records(
         :param str value_render_option: (optional) Determines how values should
             be rendered in the the output. See `ValueRenderOption`_ in
             the Sheets API.
+        :param list expected_headers: (optional) List of expected headers, they must be unique.
```
Consider adding something like: `Dictionaries will be returned for all headers, even if not included in this list.`

robrap commented on gspread/worksheet.py:
```diff
@@ -412,9 +414,27 @@ def get_all_records(
+        headers = set(keys)
+
+        # make sure they are uniques
+        if len(expected) != len(expected_headers):
+            raise GSpreadException("the given 'expected_headers' are not uniques")
+
+        if not expected & headers == expected:
+            raise GSpreadException(
+                "the given 'expected_headers' contains unknown headers: {}".format(
+                    expected & headers
+                )
+            )
```
Consider test coverage to prove the logic is and remains correct.

robrap replied:
Consider test coverage to prove the logic is and remains correct.

robrap reviewed (approved):
Thanks. This would be great. Added some comments for you to consider, but nothing blocking this.

robrap reviewed (commented):
I forgot one more comment.

lavigne958 commented:
> Thanks. This would be great. Added some comments for you to consider, but nothing blocking this.

Thank you for your review, it is much appreciated.

You're right my first test scenario was too simple, I improved it to check the new logic and added a note to mention all the headers are returned.

lavigne958 commented:
> I forgot one more comment.

you're right, this was probably an old comment and the code part was copied over and over with time. I cleaned it.

robrap commented on tests/worksheet_test.py:
```diff
@@ -629,6 +629,45 @@ def test_get_all_records_duplicate_keys(self):
+    def test_get_all_records_expected_headers(self):
+        self.sheet.resize(4, 4)
+
+        # put in new values
+        rows = [
+            ["A1", "B2", "C3", "D4"],
+            [1, "b2", 1.45, ""],
+            ["", "", "", ""],
+            ["A4", 0.4, "", 4],
+        ]
+        cell_list = self.sheet.range("A1:D4")
+        for cell, value in zip(cell_list, itertools.chain(*rows)):
+            cell.value = value
+        self.sheet.update_cells(cell_list)
```
[idea] The tests might be simpler to read if this were moved to a helper method like `update_sheet_cell_values` (or something like that).

lavigne958 replied:
you're right but this will be done during some refactoring of the test suite. with generic helper functions to generate data etc.

thank you for you feedback.

robrap reviewed (approved):
Nice. Thanks. I added one more idea to consider for the future.

robrap commented:
Also, in case this helps anyone, this is still a backward incompatible change for some. If you newly run into an issue, you’ll need to update your code to pass `expected_headers` with a list of any of the headers you expect. I’m not sure if sending an empty list would provide the legacy behavior.
